sdk/monitor/monitor-ingestion/src/utils/concurrentPoolHelper.ts

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+// Reference - https://stackoverflow.com/a/60833870/2170938
+
+export async function* asyncPool(
+  maxConcurrency: number,
+  workerArray: Array<any>,
+  iteratorFunc: (args: any) => Promise<void>
+) {
+  return new Promise((executor) => {
+    const workerQueue = [...workerArray].reverse();
+    let activeCount = 0;
+
+    const promises: Array<any> = [];
+
+    const pollNext = () => {
+      if (workerQueue.length === 0 && activeCount === 0) {
+        executor(promises);
+      } else {
+        while (activeCount < maxConcurrency && workerQueue.length) {
+          // getting the index for the current worker
+          const index = workerArray.length - workerQueue.length;
+          const worker = workerQueue.pop();
+          activeCount++;
+          const currentWorker = iteratorFunc(worker);
```

Comment:
on line 35 we do call it though. 
```ts currentWorker.then((currentWorker) => processCurrentPromise(currentWorker, index));```
The `currentWorker` is basically the promise returned from line 26.

In our context it will be the uploadCallback from logsIngestionClient.ts